versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/BinaryContent.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright (C) 2020 Dremio
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.projectnessie.versioned.persist.adapter;
+
+import com.google.protobuf.ByteString;
+import javax.annotation.Nullable;
+import org.immutables.value.Value;
+
+/** Represents all fields of a "binary object" in Nessie. */
+@Value.Immutable
+public interface BinaryContent {
+
+  /** The content ID, part of the {@link BinaryKey}, to which this binary object belongs. */
+  ContentId getContentId();
+
+  /**
+   * The object type, part of the {@link BinaryKey}, represented by this binary object, for example
+   * {@code snapshot} or {@code schema}.
+   */
+  String getObjectType();
+
+  /** The object ID, part of the {@link BinaryKey}, for example the snapshot-ID or schema-ID. */
+  String getObjectId();
+
+  /** Compression algorithm needed to uncompress {@link #getData()}. */
+  Compression getCompression();
+
+  /** Potentially compressed binary data. */
+  ByteString getData();
+
+  /** Optional "hash" value, used for optimistic locking. */
```

Comment:
Optional in what sense? Do we not control it completely on the server side?

Reply:
The issue here, actually: where I'm not yet sure about, is whether it can/should have a "proper hash" or whether some random number is sufficient. So the term "hash" is probably wrong.